### PR TITLE
Fix races and respect max stream limit when creating streams.

### LINKED ISSUE #70: Add SendStream to UnidirectionalStreamsTransport can be racy
> 2. Let stream be the newly created SendStream object.

> 3. Add stream to transport’s [[OutgoingStreams]] slot.

> 3. Continue the following steps in the background.

> 4. Create stream’s associated underlying transport.

If underlying transport is pooled Http3 connection. This can be racy. The spec checks if the max stream limit  allowed creating new streams, than creats a new SendStream and than it adds it to the trabsport. Since the transport may be used by normal browsing the limit may be already reached again.

### PR DESCRIPTION
**Pull request by jan-ivar**
Fixes #70.

***
<a href="https://pr-preview.s3.amazonaws.com/jan-ivar/webtransport/pull/202.html" title="Last updated on Feb 27, 2021, 5:28 PM UTC (e3c5627)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/webtransport/202/a813b4c...jan-ivar:e3c5627.html" title="Last updated on Feb 27, 2021, 5:28 PM UTC (e3c5627)">Diff</a>

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -207,16 +206,18 @@ interface mixin UnidirectionalStreamsTransport {
...
-1. Let |stream| be the newly created {{SendStream}} object.
-1. Add |stream| to |transport|'s {{[[OutgoingStreams]]}} slot.
-1. Continue the following steps in the background.
-1. Create |stream|'s associated underlying transport.
+1. Reserve a unidirectional stream |association| in the underlying transport.
+1. Queue a task to run the following sub-steps:
+  1. If |transport|'s state is `"closed"` or `"failed"`, abort these sub-steps.
+  1. Let |stream| be a newly created {{SendStream}} for |association|.
+  1. Add |stream| to |transport|'s {{[[OutgoingStreams]]}} slot.
+  1. resolve |p| with |stream|.
```
Comment (ricea) [inline index.bs:220]:
```suggestion
  1. Resolve |p| with |stream|.
```

File: index.bs [hunk 2]
```diff
@@ -301,15 +296,16 @@ interface mixin BidirectionalStreamsTransport {
 
 <div algorithm="add the BidirectionalStream">
```
Comment (ricea) [inline index.bs:297]:
```suggestion
<div algorithm="create a BidirectionalStream">
```

File: index.bs [hunk 3]
```diff
@@ -301,15 +296,16 @@ interface mixin BidirectionalStreamsTransport {
...
-1. Add |stream| to |transport|'s {{[[OutgoingStreams]]}} slot.
-1. Continue the following steps in the background.
-1. Create |stream|'s associated underlying transport.
+1. Reserve a bidirectional stream |association| in the underlying transport.
+1. Queue a task to run the following sub-steps:
+  1. If |transport|'s state is `"closed"` or `"failed"`, abort these sub-steps.
+  1. Let |stream| be a newly created {{BidirectionalStream}} object for |association|.
+  1. Add |stream| to |transport|'s {{[[ReceivedBidirectionalStreams]]}} slot.
+  1. Add |stream| to |transport|'s {{[[OutgoingStreams]]}} slot.
+  1. return |stream|.
```
Comment (ricea) [inline index.bs:308]:
```suggestion
  1. Resolve |p| with |stream|.
```

File: index.bs [hunk 4]
```diff
@@ -301,15 +296,16 @@ interface mixin BidirectionalStreamsTransport {
 
 <div algorithm="add the BidirectionalStream">
 
-To <dfn>add the BidirectionalStream</dfn> to a
-{{BidirectionalStreamsTransport}}, run the following steps:
+ To <dfn export for="BidirectionalStream" lt="create|creating">create</dfn> a
+ {{BidirectionalStream}} given a <var>transport</var>, run the following steps:
```
Comment (ricea) [inline index.bs:300]:
```suggestion
 {{BidirectionalStream}} given a <var>transport</var> and a promise |p|, run
  the following steps:
```

**Comment by jan-ivar** [comment]
@ricea Does this look good now?

**Review by ricea — APPROVED** [review]

**Comment by ricea** [comment]
> @ricea Does this look good now?

Yes, thanks.

**Pull request merged**